integrations/gradle-bloop/src/main/scala/bloop/integrations/gradle/model/BloopConverter.scala

```diff
@@ -399,6 +613,152 @@ class BloopConverter(parameters: BloopParameters) {
       .toMap
   }
 
+  private def getAndroidClassPathItems(
+      classPathFiles: List[File],
+      allOutputsToSourceSets: Map[File, SourceProvider],
+      allSourceSetsToProjectVariants: Map[SourceProvider, (Project, BaseVariant)],
+      targetDir: File
+  ): List[Path] = {
+    classPathFiles
+      .map(f => {
+        // change Gradle JAR references -> Bloop project classes dirs where possible.
+        allOutputsToSourceSets
+          .get(f)
+          .map(ss => {
+            val (project, variant) = allSourceSetsToProjectVariants(ss)
+            getAndroidClassesDir(targetDir, project, variant)
+          })
+          .getOrElse(f.toPath)
+      })
+      .distinct
+  }
+
+  private def getAndroidSourceSetProjectVariantMap(
+      rootProject: Project
+  ): Map[SourceProvider, (Project, BaseVariant)] = {
+    (for {
+      project <- rootProject.getAllprojects.asScala
+      nonTestVariant <- project.androidVariants
+      variant <- nonTestVariant :: Option(nonTestVariant.getTestVariant).toList
+      sourceSet <- variant.getSourceSets.asScala
+    } yield (sourceSet -> ((project, variant)))).toMap
+  }
+
+  private def getAndroidOutputsSourceSetMap(
+      allSourceSetsToProjectVariants: Map[SourceProvider, (Project, BaseVariant)]
+  ): Map[File, SourceProvider] = {
+    allSourceSetsToProjectVariants.flatMap {
+      case (sourceProvider, (project, variant)) =>
+        val bundleAppTasks = project.getTasks
+          .withType(classOf[BundleAllClasses])
+          .asScala
+          .filter(_.getVariantName == variant.getName)
+
+        val apps = bundleAppTasks
+          .flatMap(task => {
+            // v4.0.0 getJavacClasses returns DirectoryProperty.
+            // v3.4.1 getJavacClasses returns a BuildableArtifact which doesn't exists anymore and isn't handled
+            val getJavacClassesMethod = task.getClass.getMethod("getJavacClasses")
+            val classes = getJavacClassesMethod.invoke(task) match {
+              case f: DirectoryProperty => Option(f.getOrNull).map(_.getAsFile).toSet
+              case _ => Set.empty[File]
+            }
+            // v4.0.0 getOutputJar returns RegularFileProperty. v3.4.1 getOutputJar returns a File
+            val getOutputJarMethod = task.getClass.getMethod("getOutputJar")
+            val a = getOutputJarMethod.invoke(task)
+            val outputJar = a match {
+              case f: File => Set(f)
+              case f: RegularFileProperty => Option(f.getOrNull).map(_.getAsFile).toSet
+            }
+            classes ++ outputJar
+          })
+
+        // uses reflection to support multiple API versions...
+        val libs =
```

Comment:
Are we able to query the exact API version and base the implementation on that fact (use exact method etc.) instead of using an exception?

Reply:
Something like:
```scala
val bundleLibraryClass = if (SemVer.isBefore(apiVersion, "4.1.0")) Class
                .forName("com.android.build.gradle.internal.tasks.BundleLibraryClasses")
                .asInstanceOf[Class[Task]]
else Class
              .forName("com.android.build.gradle.internal.tasks.BundleLibraryClassesJar")
              .asInstanceOf[Class[Task]]
```

Are we even able to get that information and does it make sens to try it?

It's really hard to read it otherwise.

Reply:
I can see an implementation of `SemVer` in `metals/mtags/src/main/scala/scala/meta/internal/semver/SemVer.scala`

Is there a common library with that class in?  I guess I can just duplicate it for now.

Reply:
I haven't seen anything, only in sbt, where it can be used from plugins.

Reply:
I've copied it and I've simplified the reflection based on version